consensus/istanbul/backend/engine.go

```diff
@@ -443,95 +389,166 @@ func (sb *Backend) UpdateValSetDiff(chain consensus.ChainReader, header *types.H
-		log.Debug("Registry address lookup failed", "err", err)
-	} else if err != nil {
-		log.Error(err.Error())
-	}
-	// Add block rewards only if goldtoken smart contract has been initialized
-	if goldTokenAddress != nil {
-		totalBlockRewards := big.NewInt(0)
-
-		infrastructureBlockReward := big.NewInt(params.Ether)
-		governanceAddress, err := contract_comm.GetRegisteredAddress(params.GovernanceRegistryId, header, state)
-		if err == contract_errors.ErrSmartContractNotDeployed {
-			log.Debug("Registry address lookup failed", "err", err)
-		} else if err != nil {
-			log.Error(err.Error())
-		}
+	_, err = gpm.UpdateGasPriceMinimum(header, state)
+	if err == nil {
+		header.Root = state.IntermediateRoot(chain.Config().IsEIP158(header.Number))
+	}
 
-		if governanceAddress != nil {
-			state.AddBalance(*governanceAddress, infrastructureBlockReward)
-			totalBlockRewards.Add(totalBlockRewards, infrastructureBlockReward)
+	if istanbul.IsLastBlockOfEpoch(header.Number.Uint64(), sb.config.Epoch) {
+		err = sb.updateValidatorScoresAndDistributeEpochPaymentsAndRewards(header, state)
+		if err == nil {
+			header.Root = state.IntermediateRoot(chain.Config().IsEIP158(header.Number))
 		}
+	}
+
+	header.UncleHash = nilUncleHash
+
+	// Assemble and return the final block for sealing
+	return types.NewBlock(header, txs, nil, receipts, randomness), nil
+}
 
-		stakerBlockReward := big.NewInt(params.Ether)
-		lockedGoldAddress, err := contract_comm.GetRegisteredAddress(params.LockedGoldRegistryId, header, state)
-		if err == contract_errors.ErrSmartContractNotDeployed {
-			log.Debug("Registry address lookup failed", "err", err, "contract id", params.LockedGoldRegistryId)
-		} else if err != nil {
-			log.Error(err.Error())
+func (sb *Backend) updateValidatorScoresAndDistributeEpochPaymentsAndRewards(header *types.Header, state *state.StateDB) error {
+	// The validator set that signs off on the last block of the epoch is the one that we need to
+	// iterate over.
+	valSet := sb.GetValidators(big.NewInt(header.Number.Int64()-1), header.ParentHash)
+
+	err := sb.updateValidatorScores(header, state, valSet)
+	if err != nil {
+		return err
+	}
+	err = sb.distributeEpochPayments(header, state, valSet)
+	if err != nil {
+		return err
+	}
+	totalEpochRewards, err := sb.distributeEpochRewards(header, state, valSet)
+	if err != nil {
+		return err
+	}
+	return sb.increaseGoldTokenTotalSupply(header, state, totalEpochRewards)
+}
+
+func (sb *Backend) updateValidatorScores(header *types.Header, state *state.StateDB, valSet []istanbul.Validator) error {
+	for _, val := range valSet {
+		// TODO: Use actual uptime metric.
+		// 1.0 in fixidity
+		uptime := math.BigPow(10, 24)
+		log.Info("Updating validator score for address", "address", val.Address(), "uptime", uptime.String())
+		err := validators.UpdateValidatorScore(header, state, val.Address(), uptime)
+		if err != nil {
```

Comment:
Here's a case where a "partial update" would be done if there ever is a failure mid-loop.

Reply:
Great catch. Using `Snapshot` and `RevertToSnapshot` should fix this